nixos/modules/security/wrappers/wrapper.c

```diff
@@ -161,22 +162,25 @@ static int make_caps_ambient(const char *selfPath)
 
 int main(int argc, char * * argv)
 {
-    // I *think* it's safe to assume that a path from a symbolic link
-    // should safely fit within the PATH_MAX system limit. Though I'm
-    // not positive it's safe...
-    char selfPath[PATH_MAX];
-    int selfPathSize = readlink("/proc/self/exe", selfPath, sizeof(selfPath));
+    int selfExe = open("/proc/self/exe", O_PATH | O_CLOEXEC | O_NOFOLLOW);
```

Comment:
I understand why you're opening with `O_PATH | O_CLOEXEC` but I'm slightly confused by `O_NOFOLLOW` which prevents opening a symbolic link, which `/proc/self/exe` is, no?

Reply:
@ixmatus `O_NOFOLLOW` prevents *following* the symlink. So when we later `fstat` and `readlinkat` we address the symlink itself, not the target.

Reply:
From `open(2)`, the section on `O_PATH`

> If pathname is a symbolic link and the O_NOFOLLOW flag is  also  specified, then  the  call  returns  a file descriptor referring to the symbolic link.

Reply:
@shlevy ah! That was the piece I was missing. Thank you for explaining.